<a href="https://colab.research.google.com/github/BOrtizL/Proyecto-7-DataScience/blob/main/4_1_1_UDD_Proyecto_M7_Modelo_de_RandomForest_ac_BarbaraOrtiz.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

En este notebook aplicarmos el modelo de RandomForest en la data de alta confianza para tener una coparación de su rendimiento.

In [73]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [74]:
import pickle
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.decomposition import PCA
from statsmodels.stats.outliers_influence import variance_inflation_factor
import matplotlib.pyplot as plt


**Cargar los datos de alta confianza transformados**

In [75]:
# Ruta del archivo original en Google Drive
ruta_archivo_drive = "/content/drive/My Drive/Proyecto 7/raw/df_concatenado_alta_confianza.pkl"

# Cargar el DataFrame desde el archivo
with open(ruta_archivo_drive, 'rb') as f:
    df_concatenado_alta_confianza = pickle.load(f)

In [76]:
# Paso 1: Separar las características y la etiqueta
X = df_concatenado_alta_confianza.drop('gender', axis=1)
y = df_concatenado_alta_confianza['gender']

# Paso 2: Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [77]:
# Paso 3: Preprocesamiento - Escalado de características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


# Paso 4: Aplicar PCA en el conjunto de entrenamiento
pca = PCA(n_components=7)  # Número óptimo de componentes
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

In [78]:
# Paso 5: Crear y entrenar el modelo RandomForest
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train_pca, y_train)

# Paso 6: Predecir en el conjunto de prueba
y_pred = rf_model.predict(X_test_pca)

# Paso 7: Evaluar el modelo
print(classification_report(y_test, y_pred))

# Calcular la matriz de confusión
conf_matrix = confusion_matrix(y_test, y_pred)
print(conf_matrix)


              precision    recall  f1-score   support

           1       1.00      0.67      0.80         3
           2       0.67      1.00      0.80         2

    accuracy                           0.80         5
   macro avg       0.83      0.83      0.80         5
weighted avg       0.87      0.80      0.80         5

[[2 1]
 [0 2]]


**Observación:**
El modelo tiene un 80% de precisión, lo que significa que está funcionando correctamente.

A continuación la interpretación de las métricas de clasificación:
- Precision: Para la Clase 1, el modelo tiene una precisión del 100%, lo que significa que de todas las predicciones que hizo para la Clase 1, el 100% fueron correctas. Para la Clase 2, la precisión es del 67%, lo que indica que el 67% de las predicciones para la Clase 2 fueron correctas.

- Recall: La Clase 1 tiene un valor de recall del 67%, lo que significa que el 67% de las muestras de la Clase 1 fueron identificadas correctamente. Para la Clase 2, el recall es del 100%, lo que indica que todas las muestras de la Clase 2 fueron identificadas correctamente.

- F1-score: Es la media armónica entre precisión y recall.
Para ambas clases, el F1-score es del 80%.

- Accuracy: La precisión global del modelo es del 80%, es decir, el 80% de las muestras fueron clasificadas correctamente.

- Matriz de Confusión:
El modelo clasificó correctamente 2 muestras de la Clase 1 y 2 muestras de la Clase 2.
Falló en clasificar 1 muestra de la Clase 1.



# **Conclusión:**

Ahora que tenemos el resultado de desempeño de ambos modelos, es el modelo de Regresión Logistica quien tiene un excelente performance y es el modelo que se sugiere avanzar en este proyecto.